--- ../argv_from_source/argv_from_source.go	Sat Sep  7 18:45:53 2024
+++ argv_from_environment.go	Sat Sep  7 18:45:53 2024
@@ -1,15 +1,17 @@
-// Program argv_from_source - shello_world with baked-in config
+// Program argv_from_environment - argv_from_source with optional environment
+// variables
 package main
 
 /*
- * argv_from_source.go
- * shello_world, but with baked-in config
+ * argv_from_environment.go
+ * argv_from_source, but also which queries environment varibles
  * By J. Stuart McMurray
  * Created 20240901
- * Last Modified 20240902
+ * Last Modified 20240906
  */
 
 import (
+	"cmp"
 	"log"
 	"net"
 	"os"
@@ -18,11 +20,17 @@
 
 // Compile-time configurables
 var (
-	Address = "localhost:4444"
-	File    = "/etc/hosts"
+	Address       = "localhost:4444"
+	AddressEnvVar = "ALPT4ATS_ADDRESS"
+	File          = "/etc/hosts"
+	FileEnvVar    = "ALPT4ATS_FILE"
 )
 
 func main() {
+	/* Config from environment overrides compile-time config. */
+	Address = cmp.Or(os.Getenv(AddressEnvVar), Address)
+	File = cmp.Or(os.Getenv(FileEnvVar), File)
+
 	/* Connect to the server. */
 	c, err := net.Dial("tcp", Address)
 	if nil != err {
